### ISSUE TITLE
array type option cannot be set with environment var

### ISSUE THREAD

**Issue by kumar303** [yargs/yargs#821]
It looks like array type options cannot be set with an environment variable, unless this usage is incorrect? Here's how I expected it to work using **yargs@7.0.1**:

````js
// script.js :

const yargs = require('yargs');

const params = yargs
  .env('MY_SCRIPT')
  .options({
    list: {
      type: 'array',
    }
  })
  .argv;

console.log(params.list);
````

First, this worked as expected:
````
$ node script.js --list one two
[ 'one', 'two' ]
````

But using an environment variable did not work as expected:
````
$ MY_SCRIPT_LIST="one two" node script.js
[ 'one two' ]
````
I expected the same array of `['one', 'two']`.

**Comment by bcoe** [comment]
@kumar303 sorry about the slow reply; I'm triaging some tickets in prep to do some hacking on yargs (hopefully this weekend); let's get this working the way you expect.

**Comment by shubheksha** [comment]
Hey @bcoe, any update on this issue?

**Comment by addaleax** [comment]
@shubheksha It’s still an issue, and I would personally say it’s valid (i.e. this should be fixed), if possible. Do you think you might be able to look into it?

**Comment by shubheksha** [comment]
Yes, I can, if you can give me a little guidance. I tried going through the source myself but unfortunately I couldn't really figure much out.

**Comment by addaleax** [comment]
@shubheksha I am really not sure myself … it looks like you might need to modify `applyEnvVars` in https://github.com/yargs/yargs-parser and do something like [this](https://github.com/yargs/yargs-parser/blob/master/index.js#L138-L140) in there?

**Comment by gajus** [comment]
`MY_SCRIPT_LIST="one two"` should not become array `["one", "two"]` because then you will not be able to create a list of values which might contain a space.

There are three possible solutions:

1. Require that array-type parameters are set using indexed env variables in a format `${VAR_NAME}_${INDEX}`, e.g. `MY_SCRIPT_LIST_0=one,MY_SCRIPT_LIST_1=two`
2. Treat array-type parameter as JSON, e.g. require that `MY_SCRIPT_LIST` contains a valid JSON array.
3. Add a new yargs configuration that defines list delimiter (defaulting to a comma or space) and follow @kumar303's example, e.g. `YARGS_LIST_DELIMITER=,; MY_SCRIPT_LIST=one,two`

@bcoe The last of this feature is a somewhat bug, though.

**Comment by kumar303** [comment]
> Add a new yargs configuration that defines list delimiter

That seems like a good solution to me. I think it would make sense to default it to a space simply because that's how it currently works on the CLI.

**Comment by bcoe** [comment]
@kumar303 sorry for dropping this on the floor. I think my temptation would be to implement @gajus' second suggestion.

If an argument has been explicitly set to the type `array` and an argument matches `/\[.*]` I think it would be reasonable to assume that it is a JSON array, and to parse.

I would love help implementing this! I've recently created a community channel for yargs, if you feel like chatting about the feature:

http://devtoolscommunity.herokuapp.com/

**Comment by gajus** [comment]
Just in case, locally, the way I've worked around this is by dropping use of `array` altogether, using string type, and documenting that user needs to provide multiple values using a specified separator. Then just splitting the string when interpreting `argv`.

**Comment by iameli** [comment]
This _almost_ works; I'd expect it to work given the double-underscore subfield syntax:

```javascript
const yargs = require("yargs");
const argv = yargs.env("LP_").options({
  server: {
    describe: "list of servers",
    type: "array",
  },
}).argv;
console.log(argv.server);
```

```
> node yargs-test.js
LP_SERVER__0=foo LP_SERVER__1=bar node yargs-test.js
{ '0': 'foo', '1': 'bar' }
```

So, this works as a workaround with `.coerce()`:

```javascript
const yargs = require("yargs");
const argv = yargs.env("LP_").options({
  server: {
    describe: "list of servers",
    type: "array",
    coerce: (arg) => {
      if (!Array.isArray(arg)) {
        const arr = [];
        for (const [key, value] of Object.entries(arg)) {
          arr[key] = value;
        }
        return arr;
      }
      return arg;
    },
  },
}).argv;
console.log(argv.server);
```
